gpt4all-chat/src/database.cpp

```diff
@@ -1937,20 +2078,222 @@ QList<int> Database::searchEmbeddings(const std::vector<float> &query, const QLi
+{
+    float bmWeight = 0.0f;
+    if (bm25q.type == 0 /*exact*/) {
+        bmWeight = 0.9f; // the highest we give
+    } else {
+        // qlength is the length of the phrases in the query by number of distinct words
+        // ilength is the length of the natural language query by number of distinct words
+        // rlength is the number of stop words removed from the natural language query to form the query
+        float queryLengthWeight = powf(bm25q.qlength, 2) / powf(float(bm25q.ilength - bm25q.rlength), 2);
+        queryLengthWeight = qBound(0.0f, queryLengthWeight, 1.0f);
+        bmWeight = queryLengthWeight;
+        bmWeight = qBound(0.0f, bmWeight, 1.0f);
+        bmWeight = (0.25f + bmWeight * (0.75f - 0.25f));
+    }
+
+#if 0
+    qDebug()
+        << "bm25q.type"         << bm25q.type
+        << "bm25q.qlength"      << bm25q.qlength
+        << "bm25q.ilength"      << bm25q.ilength
+        << "bm25q.rlength"      << bm25q.rlength
+        << "bmWeight"           << bmWeight;
 #endif
 
-    std::vector<float> queryEmbd = m_embLLM->generateQueryEmbedding(text);
+    return bmWeight;
+}
+
+QList<int> Database::reciprocalRankFusion(const std::vector<float> &query, const QList<int> &embeddingResults,
+    const QList<int> &bm25Results, const BM25Query &bm25q, int k)
+{
+    // We default to the embedding results and augment with bm25 if any
+    QList<int> results = embeddingResults;
+
+    QList<int> missingScores;
+    QHash<int, int> bm25Ranks;
+    for (int i = 0; i < bm25Results.size(); ++i) {
+        if (!results.contains(bm25Results[i]))
+            missingScores.append(bm25Results[i]);
+        bm25Ranks[bm25Results[i]] = i + 1;
+    }
+
+    if (!missingScores.isEmpty()) {
+        QList<int> scored = scoreChunks(query, missingScores);
+        results << scored;
+    }
+
+    QHash<int, int> embeddingRanks;
+    for (int i = 0; i < results.size(); ++i)
+        embeddingRanks[results[i]] = i + 1;
+
+    const float bmWeight = bm25Results.isEmpty() ? 0 : computeBM25Weight(bm25q);
+    const int fusion_k = 60;
+
+    std::stable_sort(
+        results.begin(), results.end(),
+        [&](const int &a, const int &b) {
+            // Reciprocal Rank Fusion (RRF)
+            const int aBm25Rank = bm25Ranks.value(a, bm25Results.size() + 1);
+            const int aEmbeddingRank = embeddingRanks.value(a, embeddingResults.size() + 1);
+            Q_ASSERT(embeddingRanks.contains(a));
+
+            const int bBm25Rank = bm25Ranks.value(b, bm25Results.size() + 1);
+            const int bEmbeddingRank = embeddingRanks.value(b, embeddingResults.size() + 1);
+            Q_ASSERT(embeddingRanks.contains(b));
+
+            const float aRRFScore = bmWeight * (1.0f / (fusion_k + aBm25Rank)) + (1.f - bmWeight) * (1.0f / (fusion_k + aEmbeddingRank));
+            const float bRRFScore = bmWeight * (1.0f / (fusion_k + bBm25Rank)) + (1.f - bmWeight) * (1.0f / (fusion_k + bEmbeddingRank));
```

Comment:
the weighted average should be abstracted into its own function for readability (outside of this lambda), e.g.:
```cpp
auto mix = [](float a, float b, float f) { return a * f + b * (1.0f - f); };
```